integrationTests/state/stateTrie_test.go

```diff
@@ -1346,3 +1347,133 @@ func TestTriePruningWhenBlockIsFinal(t *testing.T) {
 	err := shardNode.AccntState.RecreateTrie(rootHashOfFirstBlock)
 	assert.Equal(t, trie.ErrHashNotFound, err)
 }
+
+func TestSnapshotOnEpochChange(t *testing.T) {
+	if testing.Short() {
+		t.Skip("this is not a short test")
+	}
+
+	numOfShards := 2
+	nodesPerShard := 3
+	numMetachainNodes := 3
+	stateCheckpointModulus := uint(3)
+
+	advertiser := integrationTests.CreateMessengerWithKadDht(context.Background(), "")
+	_ = advertiser.Bootstrap()
+
+	nodes := integrationTests.CreateNodesWithCustomStateCheckpintModulus(
+		numOfShards,
+		nodesPerShard,
+		numMetachainNodes,
+		integrationTests.GetConnectableAddress(advertiser),
+		stateCheckpointModulus,
+	)
+
+	roundsPerEpoch := uint64(5)
+	for _, node := range nodes {
+		node.EpochStartTrigger.SetRoundsPerEpoch(roundsPerEpoch)
+	}
+
+	idxProposers := make([]int, numOfShards+1)
+	for i := 0; i < numOfShards; i++ {
+		idxProposers[i] = i * nodesPerShard
+	}
+	idxProposers[numOfShards] = numOfShards * nodesPerShard
+
+	integrationTests.DisplayAndStartNodes(nodes)
+
+	defer func() {
+		_ = advertiser.Close()
+		for _, n := range nodes {
+			_ = n.Node.Stop()
+		}
+	}()
+
+	sendValue := big.NewInt(5)
+	receiverAddress := []byte("12345678901234567890123456789012")
+	initialVal := big.NewInt(10000000)
+
+	integrationTests.MintAllNodes(nodes, initialVal)
+
+	round := uint64(0)
+	nonce := uint64(0)
+	round = integrationTests.IncrementAndPrintRound(round)
+	nonce++
+
+	time.Sleep(integrationTests.StepDelay)
+
+	checkpointsRootHashes := make(map[int][][]byte)
+	snapshotsRootHashes := make(map[int][][]byte)
+	prunedRootHashes := make(map[int][][]byte)
+
+	numRounds := uint32(9)
+	for i := uint64(0); i < uint64(numRounds); i++ {
+
+		integrationTests.ProposeBlock(nodes, idxProposers, round, nonce)
+		integrationTests.SyncBlock(t, nodes, idxProposers, round)
+		round = integrationTests.IncrementAndPrintRound(round)
+		nonce++
+
+		for _, node := range nodes {
+			integrationTests.CreateAndSendTransaction(node, sendValue, receiverAddress, "")
+		}
+		time.Sleep(integrationTests.StepDelay)
+
+		for j := 0; j < numOfShards*nodesPerShard; j++ {
```

Comment:
you can also extract this loop into a function e.g collectSnapshotsAndCheckpoints

Reply:
Done.